### PR TITLE
greendns: don't contact nameservers if one entry is returned from hosts file

### PR DESCRIPTION
**Pull request by danalsan**
getaddrinfo() behaves differently from the standard implementation as
it will try to contact nameservers if only one (IPv4 or IPv6) entry
is returned from /etc/hosts file.

This patch avoids getaddrinfo() querying nameservers if at least one
entry is fetched through the hosts file to match the behavior of
the original socket.getaddrinfo() implementation.

Closes #515

Signed-off-by: Daniel Alvarez <dalvarez@redhat.com>

### CODE REVIEW

File: eventlet/support/greendns.py [hunk 1]
```diff
@@ -398,10 +399,12 @@ def getaliases(self, hostname):
 resolver = ResolverProxy(hosts_resolver=HostsResolver())
 
 
-def resolve(name, family=socket.AF_INET, raises=True, _proxy=None):
+def resolve(name, family=socket.AF_INET, raises=True, _proxy=None,
+            use_dns=True):
```
Comment (temoto) [inline eventlet/support/greendns.py]: Unless this name is commonly known in other APIs, could you please rename it to `use_network`?

File: eventlet/support/greendns.py [hunk 2]
```diff
@@ -469,16 +473,19 @@ def _getaddrinfo_lookup(host, family, flags):
...
+                        raise
+                    err = e
+                else:
+                    if answer.rrset:
+                        addrs.extend(rr.address for rr in answer.rrset)
+            if addrs:
+                break
         if err is not None and not addrs:
             raise err
     elif family == socket.AF_INET6 and flags & socket.AI_V4MAPPED:
```
Comment (temoto) [inline eventlet/support/greendns.py]: When `resolve()` family is specified, it works as before, going to network. Is that how `getaddrinfo` works also?

Comment (temoto) [reply eventlet/support/greendns.py]: There's another `else` block below too.

**Comment by temoto** [comment]
Could you write a test for new behavior?

File: eventlet/support/greendns.py (hunk 1 above)
Comment (danalsan) [reply eventlet/support/greendns.py]: ack

File: eventlet/support/greendns.py (hunk 2 above)
Comment (danalsan) [reply eventlet/support/greendns.py]: When resolve() family is specified it works as before, ie. returning the entry on hosts file and not using network if it's there.
My change basically applies to getaddrinfo() where if a IPv4 or IPv6 entry was fetched from /etc/hosts it won't attempt to get it from the network. Before this patch, if for example only an IPv4 entry was found, it'll query the network for the IPv6 resolution. Makes sense?

**Comment by danalsan** [comment]
I sent a new version with the test

File: eventlet/support/greendns.py (hunk 2 above)
Comment (temoto) [reply eventlet/support/greendns.py]: Yes, sorry, I meant `_getaddrinfo_lookup()` with family not AF_UNSPEC.

File: tests/greendns_test.py [hunk 3]
```diff
@@ -918,6 +923,57 @@ def test_hosts_priority():
...
+    addr_from_ns = '1.0.2.0'
+    addr6_from_ns = 'dead:beef::1'
+
+    hr = _make_host_resolver()
+    rp = greendns.ResolverProxy(hosts_resolver=hr, filename=None)
+    base = _make_mock_base_resolver()
+    base.rr.address = addr_from_ns
+    base.rr6.address = addr6_from_ns
+    rp._resolver = base()
+    greendns.resolver = rp
```
Comment (temoto) [inline tests/greendns_test.py]: This modifies global state for all other tests. Please use `@/with mock.patch` or isolated test or `try/finally`.

File: eventlet/support/greendns.py (hunk 2 above)
Comment (danalsan) [reply eventlet/support/greendns.py]: Yes if family not AF_UNSPEC then my patch is not changing the behavior

Comment (temoto) [reply eventlet/support/greendns.py]: It should, to match system `getaddrinfo`, right?

Comment (danalsan) [reply eventlet/support/greendns.py]: @temoto before my patch, individual calls to `resolve()` (ie. family not being AF_UNSPEC) where already checking hosts file first. Problem was that calls with AF_UNSPEC will call twice to resolve (one with AF_INET6 and one with AF_INET) and if only one (4 or 6) entry is present in hosts file, the other call will contact the network.

**Comment by temoto** [comment]
@jstasiak review please

**Comment by mangelajo** [comment]
Looks good to me! :)

**Comment by temoto** [comment]
@danalsan thank you, merged in 10ff67f84a1e2f53c8eb254944ebf81762041b45

**Pull request merged**